### DISCUSSION TITLE
How to switch in recovery shell when rollback is not possible

### DISCUSSION THREAD

**Discussion by 573** [nix-community/NixOS-WSL#340] [Q&A]
I'm in the delicate position to have my WSL rendered unable to boot my NixOS-WSL VM since yesterdays Windows-Updates (search machine: `Error code: Wsl/ERROR_NOT_FOUND`):
> wsl -v
WSL-Version: 2.0.9.0
Kernelversion: 5.15.133.1-1
WSLg-Version: 1.0.59
MSRDC-Version: 1.2.4677
Direct3D-Version: 1.611.1-81528511
DXCore-Version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows-Version: 10.0.19045.3570

Edited: The following is old information, currently I only have the Microsoft-Store version installed and unusable. To run the following I earlier did install the non-Store version (i. e. uninstalling Store-version via Apps & Features first and then doing `wsl -update` if needed and then installing that one https://learn.microsoft.com/de-de/windows/wsl/install-manual#step-4---download-the-linux-kernel-update-package) to run at least a bit of recovery / backup of my VM's data, see [here if interested](https://gist.github.com/573/3f6152e28a4abfc6631ae0faa3ae9395). \*rant TBH To me it seems as if M$ does give AF to consumers and just wants them to use their paid appliances or whatever \*rantover and also ❤️ to this awesome project here.

For now the standard Ubuntu VM is booting again. So I decided to trim down my Nixos-WSL a bit and prepare my flakes-based config for reinstallation as follows on a second VM, random list as specs are rather rare: no nativeSystemd, no interop, no wsl-vpnkit.

Using the Debugging tipps from our issues list I came so far as to run the recovery shell and therein spawn more or less the command:

>nixos-rebuild switch --option accept-flake-config true --flake .#MYHOSTNAME

But to the result that I recognised `systemd` is not running in the recovery shell so I cannot `switch`. Latter giving

> error: getting attributes of path '/run/binfmt': No such file or directory

Indeed:
>[nix-shell:~/nix-config-1]# systemctl status binfmt
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down

How to proceed ?

**Comment by 573** [comment]
Not an answer to the question but in case someone needs to know this:

# Get back to normal after Windows-Update disruption - Issue on the other repo
The steps I took where re-installing the Windows-Store version (so basically getting all three icons above in the famous screen above) after I took this exact measure:

> Thank you @pahar0. This confirms our theory: There is an old WSL installation that's conflicting with the latest package.
> 
> Can you try to:
> 
>     * Delete: `HKEY_LOCAL_MACHINE\SOFTWARE\Classes\PackagedCom\ClassIndex\{A9B7A1B9-0671-405C-95F1-E0612CB4CE7E}`
> 
>     * See if that solves the issue
> 
> 
> (You can use the .reg you just captured as backup if needed)

_Originally posted by @OneBlue in https://github.com/microsoft/WSL/issues/10782#issuecomment-1820066209_
            
Then all seems back to normal:
> wsl -v
WSL-Version: 2.0.9.0
Kernelversion: 5.15.133.1-1
WSLg-Version: 1.0.59
MSRDC-Version: 1.2.4677
Direct3D-Version: 1.611.1-81528511
DXCore-Version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows-Version: 10.0.19045.3570

wsl -d NixOS-Groundtruth -- /run/current-system/sw/bin/uname -sr
Linux 5.15.133.1-microsoft-standard-WSL2

_Originally posted by @573 in https://github.com/microsoft/WSL/issues/10767#issuecomment-1821652246_

**Comment by nzbr** [comment]
If you can not switch due to systemd not running, you should still be able to rebuild the systemd by using boot instead. Then just restart the distro or VM with either `wsl -t <distro>` or `wsl --shutdown`

Does NixOS work normally when you try to import the latest release's `nixos-wsl.tar`?
